### PR TITLE
Fix the handling of removing disconnected streams to avoid a panic when multiple streams disconnect from the sdkserver

### LINKED ISSUE #3640: runtime.boundsError from watchGameServer
**What happened**:

Sidecar process crashes

**What you expected to happen**:

Sidecar process never crashes, regardless of input

**How to reproduce it (as minimally and precisely as possible)**:

I am interacting with agones using the NodeJS sdk. I am reusing the same gameserver across multiple restarts. My game server code exits, gets restarted then connects to the same gameserver again. This game server might still be allocated from the last finished game, so it needs to be set to ready again. Let me know if there is a better way to ensure that the gameserver is in a ready state.

```
    // wait for ready state to be committed before continuing to premature allocation where the state of the previous server gets carried over and server gets instantly allocated
    // awaiting ready() doesn't necessarily mean the state has been committed, watchGameServer needs to report ready
    // https://agones.dev/site/docs/guides/client-sdks/#function-reference
    const clientSDK = this.clientSDK;
    const readyPromise = new Promise<void>((resolve) => {
      clientSDK.watchGameServer(
        (g) => {
          if (g.status.state === "Ready") resolve();
        },
        () => {}
      );
    });
    await clientSDK.ready();
    await readyPromise;
```

**Anything else we need to know?**:
n/a

**Environment**:
- Agones version: 1.36.0
- Kubernetes version (use `kubectl version`):
Client Version: v1.29.0
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.28.5-eks-5e0fdde
- Cloud provider or hardware configuration: AWS, Amazon Linux 2
- Install method (yaml/helm): helm
- Troubleshooting guide log(s):
```
{"ctlConf":{"GameServerName":"medium-4bs9s-zgb8b","PodNamespace":"servers","Address":"localhost","IsLocal":false,"LocalFile":"","Delay":0,"Timeout":0,"Test":"","TestSdkName":"","KubeConfig":"","GracefulTermination":true,"GRPCPort":9357,"HTTPPort":9358},"featureGates":"CountsAndLists=false\u0026Example=false\u0026FleetAllocationOverflow=false\u0026PlayerAllocationFilter=false\u0026PlayerTracking=false\u0026PodHostname=true\u0026ResetMetricsOnDelete=true\u0026SplitControllerAndExtensions=true","message":"Starting sdk sidecar","severity":"info","source":"main","time":"2024-02-09T18:34:19.434921673Z","version":"1.36.0"}
W0209 18:34:19.435123       1 client_config.go:618] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
{"gsKey":"servers/medium-4bs9s-zgb8b","message":"Created GameServer sidecar","severity":"info","source":"*sdkserver.SDKServer","time":"2024-02-09T18:34:19.435485156Z"}
{"gsKey":"servers/medium-4bs9s-zgb8b","message":"Connection to Kubernetes service established","severity":"info","source":"*sdkserver.SDKServer","time":"2024-02-09T18:34:19.446843136Z","try":0}
{"grpcEndpoint":"localhost:9357","message":"Starting SDKServer grpc service...","severity":"info","source":"main","time":"2024-02-09T18:34:19.448284832Z"}
{"httpEndpoint":"localhost:9358","message":"Starting SDKServer grpc-gateway...","severity":"info","source":"main","time":"2024-02-09T18:34:19.448979335Z"}
{"gsKey":"servers/medium-4bs9s-zgb8b","message":"Starting workers...","queue":"agones.dev.servers.medium-4bs9s-zgb8b","severity":"info","source":"*sdkserver.SDKServer","time":"2024-02-09T18:34:19.548261609Z","workers":1}
{"error":"context canceled","gsKey":"servers/medium-4bs9s-zgb8b","message":"stream closed with error","severity":"error","source":"*sdkserver.SDKServer","time":"2024-02-09T18:48:24.742620632Z"}
{"error":"context canceled","gsKey":"servers/medium-4bs9s-zgb8b","message":"stream closed with error","severity":"error","source":"*sdkserver.SDKServer","time":"2024-02-09T18:48:24.778610652Z"}
{"error":"context canceled","gsKey":"servers/medium-4bs9s-zgb8b","message":"stream closed with error","severity":"error","source":"*sdkserver.SDKServer","time":"2024-02-09T19:07:49.493424215Z"}
E0209 19:07:49.494045       1 runtime.go:79] Observed a panic: runtime.boundsError{x:2, y:1, signed:true, code:0x3} (runtime error: slice bounds out of range [2:1])
goroutine 71 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x1d56400?, 0xc0006f2528})
	/go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:75 +0x99
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc0003642c0?})
	/go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:49 +0x75
panic({0x1d56400, 0xc0006f2528})
	/usr/local/go/src/runtime/panic.go:884 +0x213
agones.dev/agones/pkg/sdkserver.(*SDKServer).sendGameServerUpdate(0xc0003641e0, 0xc00017ec00)
	/go/src/agones.dev/agones/pkg/sdkserver/sdkserver.go:1248 +0x7da
agones.dev/agones/pkg/sdkserver.NewSDKServer.func2({0x406ad8?, 0xc00010d2c0?}, {0x1e75f80?, 0xc00017ec00?})
	/go/src/agones.dev/agones/pkg/sdkserver/sdkserver.go:184 +0x36
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate(...)
	/go/src/agones.dev/agones/vendor/k8s.io/client-go/tools/cache/controller.go:250
k8s.io/client-go/tools/cache.(*processorListener).run.func1()
	/go/src/agones.dev/agones/vendor/k8s.io/client-go/tools/cache/shared_informer.go:971 +0xfc
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)
	/go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x3e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc000583f38?, {0x22df9c0, 0xc0005ad380}, 0x1, 0xc00010dda0)
	/go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x405d3d?, 0x3b9aca00, 0x0, 0x0?, 0xc000583f88?)
	/go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:204 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:161
k8s.io/client-go/tools/cache.(*processorListener).run(0xc000594bd0)
	/go/src/agones.dev/agones/vendor/k8s.io/client-go/tools/cache/shared_informer.go:967 +0x6b
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
	/go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:72 +0x5a
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start
	/go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:70 +0x85
panic: runtime error: slice bounds out of range [2:1] [recovered]
	panic: runtime error: slice bounds out of range [2:1]

goroutine 71 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc0003642c0?})
	/go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:56 +0xd7
panic({0x1d56400, 0xc0006f2528})
	/usr/local/go/src/runtime/panic.go:884 +0x213
agones.dev/agones/pkg/sdkserver.(*SDKServer).sendGameServerUpdate(0xc0003641e0, 0xc00017ec00)
	/go/src/agones.dev/agones/pkg/sdkserver/sdkserver.go:1248 +0x7da
agones.dev/agones/pkg/sdkserver.NewSDKServer.func2({0x406ad8?, 0xc00010d2c0?}, {0x1e75f80?, 0xc00017ec00?})
	/go/src/agones.dev/agones/pkg/sdkserver/sdkserver.go:184 +0x36
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate(...)
	/go/src/agones.dev/agones/vendor/k8s.io/client-go/tools/cache/controller.go:250
k8s.io/client-go/tools/cache.(*processorListener).run.func1()
	/go/src/agones.dev/agones/vendor/k8s.io/client-go/tools/cache/shared_informer.go:971 +0xfc
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)
	/go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x3e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc000583f38?, {0x22df9c0, 0xc0005ad380}, 0x1, 0xc00010dda0)
	/go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x405d3d?, 0x3b9aca00, 0x0, 0x0?, 0xc000583f88?)
	/go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:204 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:161
k8s.io/client-go/tools/cache.(*processorListener).run(0xc000594bd0)
	/go/src/agones.dev/agones/vendor/k8s.io/client-go/tools/cache/shared_informer.go:967 +0x6b
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
	/go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:72 +0x5a
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start
	/go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:70 +0x85
```

### PR DESCRIPTION
**Pull request by roberthbailey**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, press enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind breaking

/kind bug

> /kind cleanup
> /kind documentation
> /kind feature
> /kind hotfix
> /kind release

**What this PR does / Why we need it**: Fixes a runtime panic when multiple streams disconnect from the sdkserver. This happens when a game server creates more than one watch and then closes them all at the same time. Normally, watches last for the lifetime of a game server, but if a game server intentionally disconnects watches and/or expects to be able to exit and restart (which implicitly disconnects the watches) then we hit this edge case. 

**Which issue(s) this PR fixes**:

Closes #3640

**Special notes for your reviewer**: I tested this fix two ways:

1. The new unit test fails with the old code paths with the error:
```
--- FAIL: TestSDKServer_SendGameServerUpdateRemovesDisconnectedStream (2.01s)
panic: runtime error: slice bounds out of range [2:1] [recovered]
	panic: runtime error: slice bounds out of range [2:1]
```
2. I installed the fix into a test cluster and verified that the nodejs sample program that triggered the panic previously no longer does so.

### CODE REVIEW

File: pkg/sdkserver/sdkserver.go [hunk 1]
```diff
@@ -1243,27 +1243,35 @@ func (s *SDKServer) sendGameServerUpdate(gs *agonesv1.GameServer) {
 	s.streamMutex.Lock()
 	defer s.streamMutex.Unlock()
 
-	for i, stream := range s.connectedStreams {
+	// Filter the slice of streams sharing the same backing array and capacity as the original
+	// so that storage is reused and no memory allocations are made. This modifies the original
+	// slice.
+	//
+	// See https://go.dev/wiki/SliceTricks#filtering-without-allocating
+	remainingStreams := s.connectedStreams[:0]
```
Comment (zmerlynn) [inline pkg/sdkserver/sdkserver.go:1251]: Ok, I get what you're doing, but do we really care? If you have to document it this thoroughly, are we sure it's worth the optimization?

**Review by zmerlynn — APPROVED** [review]
I'm ok either way with the below comment - leave it if you wish. :)

File: pkg/sdkserver/sdkserver.go (hunk 1 above)
Comment (roberthbailey) [reply pkg/sdkserver/sdkserver.go:1251]: I believe that this loop gets executed quite frequently (every time we send a game server change event, so when game servers move to ready, allocated, shutdown, etc or have labels set, counts or lists modified, etc.). 

Code-wise, it isn't much different from allocating a new slice, appending to it, and then replacing the `connectedStreams` (the only different being line 1251 which would change to `remainingStreams := make(...)`. So making it more memory efficient doesn't seem to have any downsides other than this extra long comment.
